### ISSUE TITLE
Wrong operator precedence between "=" (assignment) and "and" (binary operator)

### ISSUE THREAD

**Issue by jj101k** [glayzzle/php-parser#81]
Given code:

```
<?php
class Foo {
    public function bar() {
        return new Bar();
    }
}
class Bar {
    public function baz() {
        return 1;
    }
}
$foo = new Foo();

if(
    $foo and
    $bar = $foo->bar() and
    $bar->baz()
) {
echo 1;
}
```

php-parser considers the if statement to be `$foo and $bar = ($foo->bar() and $bar->baz())`, which would be an error because `$bar` isn't defined before it's used. Real PHP considers it to be `$foo and ($bar = $foo->bar()) and $bar->baz()`, which works fine.

You can do a simple comparison by running:

`node -e 'var fs = require("fs"); var phpParser = require("php-parser"); var parser = new phpParser({}); console.log(parser.parseCode(fs.readFileSync("test.php", "utf8")).children[3])'`

Versus

`php test.php`

This pattern is fairly common, analogous to "optional chaining", although normally people have the sense to bracket for clear precedence.

http://php.net/manual/en/language.operators.precedence.php

Assignment is above "and" so it binds tighter.

**Note** "and" does not have the same precedence as "&&". For historical reasons, `a and b;` as a statement works like a shorthand for `if(a) {b;}`, so "and" has to be as low-precedence as possible.

**Comment by ichiriac** [comment]
related to https://github.com/glayzzle/php-parser/issues/75

**Comment by ichiriac** [comment]
fixed & released under 2.0.5

**Comment by jj101k** [comment]
I'm still getting the same behaviour on version 2.0.5.

```
$ git describe --tags HEAD
2.0.4-13-g2372443
$ node -e 'var fs = require("fs"); var phpParser = require("./src/index.js"); var parser = new phpParser({}); console.log(parser.parseCode(fs.readFileSync("test.php", "utf8")).children[3])'
If {
  kind: 'if',
  test: 
   Bin {
     kind: 'bin',
     type: 'and',
     left: Variable { kind: 'variable', name: 'foo', byref: false, curly: false },
     right: Assign { kind: 'assign', operator: '=', left: [Object], right: [Object] } },
  body: Block { kind: 'block', children: [ [Object] ] },
  alternate: null,
  shortForm: false }
$ 
```

**Comment by ichiriac** [comment]
hi @jj101k ,

I've forget to try your issue, now it's fixed and released in 2.0.6

**Comment by jj101k** [comment]
Thanks. Confirmed that it has the correct structure in 2.0.6
